[wxQt] Fix  wxEVT_TREE_SEL_{CHANGING,CHANGED} generation/handling

AliKet opened the pull request.

Commit:
wxQt: Remove trailing white-space

No real changes.

Commit:
wxQt: Simplify wxTreeCtrl::GetSelection() implementation

This also fixes returning the correct wxTreeItemId when this function is called
from a wxEVT_TREE_SEL_CHANGED handler.

Notice that this function is just a synonym for GetFocusedItem() now for controls
without wxTR_MULTIPLE style.

Commit:
wxQt: Fix generating wxEVT_TREE_SEL_{CHANGING,CHANGED} events

The wxEVT_TREE_SEL_CHANGING event is now generated before currentItemChanged()
signal is emitted so we will have the opportunity to not emit the signal altogether
if the wx event is vetoed.

The wxEVT_TREE_SEL_CHANGED event is now generated when the last itemSelectionChanged()
signal is emitted, in which case wxTreeCtrl::GetSelections() is guaranteed to return
the correct selected items if called from the event handler.

Also fix a bug when wxTreeCtrl::EndEditLabel() is called in a control with wxTR_MULTIPLE
flag resulting in GetSelections() returning the old selected item instead of the new one.

And remove a redundant call to setEditTriggers().

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -125,6 +125,61 @@ class ImageState
+namespace
+{
+// Used when generating wxEVT_TREE_SEL_{CHANGING,CHANGED} events only
+static wxTreeItemId gs_currentItem;
+static wxTreeItemId gs_previousItem;
```
Do these really need to be global? I am not 100% sure it's not going to be a problem, e.g. if a handler of some event changes selection in some other tree control.

vadz commented on src/qt/treectrl.cpp:
```diff
@@ -125,6 +125,61 @@ class ImageState
 Q_DECLARE_METATYPE(TreeItemDataQt)
 
+// We use a custom selection model because we need to generate the
+// wxEVT_TREE_SEL_CHANGED event only if it is allowed. that is, only
```
Nitpick of the day:
```suggestion
// wxEVT_TREE_SEL_CHANGED event only if it is allowed, that is, only
```

(alternatively, keep the period, but capitalize "That")

AliKet replied:
forgot to update this

vadz reviewed (commented):
Thanks for the fixes, I think the logic is correct here, but if we could store the items in, say, `wxQTreeWidget`, as member variables, it would be better IMHO.

AliKet commented:
@dsa-t It would be really apreciated if you could test this PR (or this [patch](https://patch-diff.githubusercontent.com/raw/wxWidgets/wxWidgets/pull/24600.diff)) before it get merged because I had to change the implementation a little due to problems encountered with `wxTreeBook` control with the previous version.

TIA.

AliKet commented:
Oops! this version reintroduced the infinite loop reported by @dsa-t which can be seen with the notebook sample (File -> Type -> Treebook)

AliKet commented:
This is really strange because the treectrl sample works correctly but the notebook (treebook page) does not! maybe the problem is in the `wxTreeBook` control !?

Needs more debugging...

AliKet commented:
> This is really strange because the treectrl sample works correctly but the notebook (treebook page) does not! maybe the problem is in the `wxTreeBook` control !?
> 
> Needs more debugging...

Everything seems to work correctly now: the treectrl and notebook (wxTreebook) samples.

dsa-t commented:
Works good generally, but I also see a small issue with multiple selections:

1. In `wxTR_MULTIPLE` mode, select `File child 1.1` to `File child 1.5`
2. Then, select, for example, `File child 1.2`
3. Inside `OnSelChanged`, `GetSelections` will say "File child 1.1""File child 1.2""File child 1.3""File child 1.4""File child 1.5"
Also the name editor gets triggered (I don't think it should).

AliKet commented:
> Works good generally, but I also see a small issue with multiple selections in `treetest`:
> 
> 1. In `wxTR_MULTIPLE` mode, select `File child 1.1` to `File child 1.5`
> 2. Then, select, for example, `File child 1.2`
> 3. Inside `OnSelChanged`, `GetSelections` will say "File child 1.1""File child 1.2""File child 1.3""File child 1.4""File child 1.5"
>    Also the name editor gets triggered (I don't think it should).

Thanks for testing, but it works correctly for me!

![Screenshot](https://github.com/wxWidgets/wxWidgets/assets/7704771/c8e54b56-a810-44ac-93d7-7bdce957a210)

dsa-t commented:
> Thanks for testing, but it works correctly for me!

In the 2nd step, I mean select just the 1.2 item, not deselect it:

![image](https://github.com/wxWidgets/wxWidgets/assets/37658952/22504f3d-8ef9-455b-931f-647b1caa3b04)

AliKet commented:
> > Thanks for testing, but it works correctly for me!
> 
> In the 2nd step, I mean select just the 1.2 item, not deselect it:

Indeed, I can reproduce the problem.  And the fix is simple which I'll push soon.
but preventing the editor from being triggered is trickier because this is how Qt interpret `SelectedClicked` flag passed here https://github.com/wxWidgets/wxWidgets/blob/f8043a62f3319719cbd3a400ea454e4dbb3d23ff/src/qt/treectrl.cpp#L1311-L1315 

i.e. It starts editing when clicking on an already selected item

See Qt docs https://doc.qt.io/qt-6/qabstractitemview.html#EditTrigger-enum:~:text=QAbstractItemView%3A%3ASelectedClicked

AliKet commented:
@dsa-t  this patch:
```C++
diff --git a/src/qt/treectrl.cpp b/src/qt/treectrl.cpp
index ea68681ae0..45be7b0c48 100644
--- a/src/qt/treectrl.cpp
+++ b/src/qt/treectrl.cpp
@@ -330,11 +330,31 @@ protected:
         }
     }
 
+    // For compatibility with MSW, selecting an item that is part of a multiple selection
+    // simply deselects the other items without triggering the edit mode on the item. This
+    // is not the case with Qt because we already tell it to start editing when clicking
+    // on an already selected item with setEditTriggers(QTreeWidget::SelectedClicked) call.
+    bool m_blockEditing = false;
+
+    void selectionChanged(const QItemSelection& selected,
+                          const QItemSelection& deselected) override
+    {
+        m_blockEditing = selected.size() == 0;
+
+        QTreeView::selectionChanged(selected, deselected);
+    }
+
     bool edit(const QModelIndex &index, EditTrigger trigger, QEvent *event) override
     {
         // AllEditTriggers means that editor is about to open, not waiting for double click
         if (trigger == AllEditTriggers)
         {
+            if ( m_blockEditing )
+            {
+                m_blockEditing = false;
+                return false;
+            }
+
             // Allow event handlers to veto opening the editor
             wxTreeEvent wx_event(
                 wxEVT_TREE_BEGIN_LABEL_EDIT,

```
solves the problem with starting edit you mention, but I'm still not sure if we should apply it or not because wxGTK3 also start editing if you do the following:
- Select `File child 1.1`
- Hold `Ctrl` key down and select `File child 1.2`
- Now you have 2 item selected (`File child 1.1` and `File child 1.2`)
- Select `File child 1.2` with or without holding the `Ctrl` key down.
- The control starts editing the item.

Notice that this happens only if you select the focused item. `wxMSW` doesn't behave like this. i.e. it doesn't start editing at all.

@vadz any thoughts about this ?

vadz commented:
I think wxGTK (i.e. generic implementation) is wrong here, this doesn't seem like a useful behaviour to me.

Commit:
Get rid of CppUnit boilerplate in wxTreeCtrl unit tests

Exclude UI tests dependent on wxUIActionSimulator (i.e. those added with
WXUISIM_TEST), they were previously ignored under wxOSX anyway.

Also, exclude only LabelEdit testcase instead of the entire TreeCtrlTestCase
when running under Wine.

Commit:
Put all WXUISIM_TEST under wxUSE_UIACTIONSIMULATOR in wxTreeCtrl unit tests

No real changes.

Commit:
Test vetoing of wxEVT_TREE_SEL_CHANGING in wxTreeCtrl unit tests

Commit:
Added SelectItemMultiInteractive test case to TreeCtrlTestCase

Selecting multiple items (in wxTreeCtrl with wxTR_MULTIPLE flag) and then
deselecting them by clicking on the current item that is also part of the
selection should deselect everything but the current item without starting
to edit that item.

Commit:
Fix unwanted item edit in wxQt wxTreeCtrl with wxTR_MULTIPLE

Commit:
Fix unwanted item edit in wxGenericTreeCtrl with wxTR_MULTIPLE

Commit:
wxQt: respect wxTR_HIDE_ROOT style flag for wxTreeCtrl

AliKet commented:
@vadz  I have no idea what's going on here, or what to do with these failures?

vadz commented:
I'm afraid that at least the Mac failures are real ones, i.e. actually due to the changes here as they seem to be occurring in the affected code. For the MSW ones I'm less sure and I don't have time to debug this right now, sorry...

AliKet commented:
@vadz This is ready now. just needs **AppVeyor** restart, TIA!

vadz commented on tests/controls/treectrltest.cpp:
```diff
@@ -25,82 +25,55 @@
+    ~TreeCtrlTestCase()
+    {
+        delete m_tree;
+        m_tree = nullptr;
+
+        m_root =
+        m_child1 =
+        m_child2 =
+        m_grandchild = wxTreeItemId();
```
Minor, but this is not really needed in a dtor.

vadz commented on tests/controls/treectrltest.cpp:
```diff
@@ -111,134 +84,103 @@ class TreeCtrlTestCase : public CppUnit::TestCase
-    delete m_tree;
-    m_tree = nullptr;
+#ifdef __WXMSW__
+    // Toggling the wxTR_HIDE_ROOT window style has no effect under wxMSW if
```
It would be nice to mention this in `wxTR_HIDE_ROOT` docs.

AliKet replied:
Oops! I forgot to update the documentation. Please could you do this when merging?

TIA!

vadz commented on tests/controls/treectrltest.cpp:
```diff
@@ -111,134 +84,103 @@ class TreeCtrlTestCase : public CppUnit::TestCase
-{
-    delete m_tree;
-    m_tree = nullptr;
+#ifdef __WXMSW__
```
I'm not sure if it's worth doing, but we could get rid of this `#ifdef` by defining `TreeCtrlHideRootTestCase` deriving from `TreeCtrlTestCase` and passing `wxTR_HIDE_ROOT` as (optional) style argument to the base class ctor.

AliKet replied:
Thanks for the suggestion.

vadz commented on .github/workflows/ci_mac.yml:
```diff
@@ -232,7 +232,7 @@ jobs:
         export ASAN_OPTIONS='fast_unwind_on_malloc=0 detect_container_overflow=0'
         # Exclude tests that are currently known to fail
-        wx_tests_selection='~[.] ~[valnum] ~wxPersistTLW ~wxPersistDVC ~wxTopLevel::Show ~wxFont::Weight ~wxFont::NativeFontInfo ~WebView ~RadioButton::Single ~RadioButton::Focus ~wxHtmlPrintout::Pagination ~wxTextCtrl::GetBestSize ~TextCtrlTestCase ~wxExecute::RedirectUTF8 ~Ellipsization::NormalCase ~SpinCtrl::* ~SpinCtrlDouble::* ~NotebookTestCase ~SettingsTestCase ~Window::Show ~ModalDialogsTe ...
+        wx_tests_selection='~[.] ~[valnum] ~wxPersistTLW ~wxPersistDVC ~wxTopLevel::Show ~wxFont::Weight ~wxFont::NativeFontInfo ~WebView ~RadioButton::Single ~RadioButton::Focus ~wxHtmlPrintout::Pagination ~wxTextCtrl::GetBestSize ~TextCtrlTestCase ~wxExecute::RedirectUTF8 ~Ellipsization::NormalCase ~SpinCtrl::* ~SpinCtrlDouble::* ~NotebookTestCase ~SettingsTestCase ~Window::Show ~ModalDialogsTe ...
```
Are there any actually tests passing under Mac? It's not obvious why we don't exclude `TreeCtrlTestCase` entirely...

AliKet replied:
Excluded tests (`wxTreeCtrl::*`) are only those that require `wxUIActionSimulator` to run. Excluding `TreeCtrlTestCase` entirely would exclude the others: `wxTreeCtrl::{HasChildren, GetCount, SelectItemSingle, SelectItemMulti, ...}` to name a few that run successfully.

vadz reviewed (commented):
Thanks, I don't know what's going on with AppVeyor and why it considers this PR unmergeable. Do you plan to rebase it again? If so, it might be enough to force it to run the builds again... If not, let me know and I'll just merge it.
